### PR TITLE
fix(router): should run resolvers for the same route concurrently

### LINKED ISSUE #14279: Run multiple resolvers concurrently in router.
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
I set multiple resolvers in my route:
```typescript
export const routes: Routes = [
  {
    path: '',
    children: appRoutes,
    canActivate: [AuthGuard],
    resolve: {
      capabilities: CapabilitiesResolver,
      userInfo: OrganizationUserResolver,
      organizations: OrganizationResolver
    }
  }
]
```
each of them implements Resolve interface and return Observable:
```typescript
@Injectable()
export class CapabilitiesResolver implements Resolve<Endpoint[]> {
  constructor(private capabilitiesService: CapabilitiesService) {}

  resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<Endpoint[]> {
    return this.capabilitiesService.fetch()
  }
}
```
as result all http requests run sequentially and this is not useful behavior for UX. But if resolvers return Promise instead of Observable:
```typescript
resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Promise<Endpoint[]> {
    return this.capabilitiesService.fetch().toPromise()
  }
```
then all requests will be ran concurrently, very implicit behavior...

**Expected behavior**
No matter what `resolve()` method return (Observable or Promise) all async operations should be ran concurrently.

**What is the motivation / use case for changing the behavior?**
For users: Improve UX when router pre-fetch data.
For developers: More explicit router API.

**Please tell us about your environment:**
@angular/cli: 1.0.0-beta.30
node: 6.2.2
os: darwin x64
@angular/common: 2.4.1
@angular/compiler: 2.4.1
@angular/core: 2.4.1
@angular/forms: 2.4.1
@angular/http: 2.4.1
@angular/platform-browser: 2.4.1
@angular/platform-browser-dynamic: 2.4.1
@angular/router: 3.4.1
@angular/compiler-cli: 2.4.1

* **Angular version:** 2.4.1

* **Browser:** all

 
* **Language:** TypeScript 2.0.3

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Fixes #14279

### CODE REVIEW

File: packages/router/test/integration.spec.ts [hunk 1]
```diff
@@ -844,24 +844,41 @@ describe('Integration', () => {
...
+      log = [];
       TestBed.configureTestingModule({
         providers: [
           {provide: 'resolveTwo', useValue: (a: any, b: any) => 2},
           {provide: 'resolveFour', useValue: (a: any, b: any) => 4},
           {provide: 'resolveSix', useClass: ResolveSix},
           {provide: 'resolveError', useValue: (a: any, b: any) => Promise.reject('error')},
-          {provide: 'numberOfUrlSegments', useValue: (a: any, b: any) => a.url.length}
+          {provide: 'numberOfUrlSegments', useValue: (a: any, b: any) => a.url.length},
+          {
```
Comment (vicb) [inline packages/router/test/integration.spec.ts]: 2 resolvers should be enough.
May be the second should complete the first and then we do not have to play with timings.

**Review by vicb — DISMISSED** [review]
Please address the comments.

Also waitForMap is called twice in the code base. Could it cause concurrency issue when loading child config (ie if both use the loadChildren) ?

**Comment by DzmitryShylovich** [comment]
> Also waitForMap is called twice in the code base. Could it cause concurrency issue when loading child config (ie if both use the loadChildren) ?

probably yes, so I decided not to use it at all. It's more readable without `waitForMap` + faster (we don't need to loop 2 times).

> May be the second should complete the first and then we do not have to play with timings.

Don't think it's a good idea. Consider such test but without the fix (resolvers run sequentially). `resolver1` will never complete because `resolver2` will never run (deadlock). So to fail such test we will need to introduce a timeout. Plus we need to add `the second should complete the first` logic. Sounds like overhead to me.

File: packages/router/test/integration.spec.ts (hunk 1 above)
Comment (DzmitryShylovich) [inline packages/router/test/integration.spec.ts]: removed the 3rd resolver

**Comment by vicb** [comment]
> So to fail such test we will need to introduce a timeout.

Jasmine will timeout by default (you can override the default timeout with a shorter one).

**Comment by vicb** [comment]
> the second should complete the first

should be trivial ?

**Comment by DzmitryShylovich** [comment]
@vicb updated. it's a little bit more code but it's sync now 😃

**Review by jasonaden — APPROVED** [review]
LGTM

**Pull request merged**